### PR TITLE
Improvements to shutdown behavior

### PR DESCRIPTION
**Pull request by ajrhacker**
- Add a configurable delay before shutting down an emulated machine to give its components fair warning of impending power failure using a new device callback. The default delay is practically unnoticeable, and the feature will potentially improve NVRAM handling in many drivers.
- UI Quit and UI Force Quit inputs are now distinguished from UI Cancel. The new "force quit" option effectively disables both the shutdown delay and -nvram_save.

### CODE REVIEW

File: src/emu/inpttype.h [hunk 1]
```diff
@@ -851,6 +851,8 @@ inline void construct_core_types_UI(simple_list<input_type_entry> &typelist)
 	INPUT_PORT_DIGITAL_TYPE( 0, UI,      UI_FOCUS_PREV,       "UI Focus Previous",      input_seq(KEYCODE_TAB, KEYCODE_LSHIFT) )
 	INPUT_PORT_DIGITAL_TYPE( 0, UI,      UI_SELECT,           "UI Select",              input_seq(KEYCODE_ENTER, input_seq::or_code, JOYCODE_BUTTON1_INDEXED(0), input_seq::or_code, KEYCODE_ENTER_PAD) )
 	INPUT_PORT_DIGITAL_TYPE( 0, UI,      UI_CANCEL,           "UI Cancel",              input_seq(KEYCODE_ESC) )
+	INPUT_PORT_DIGITAL_TYPE( 0, UI,      UI_QUIT,             "UI Quit",                input_seq(KEYCODE_ESC, input_seq::not_code, KEYCODE_LSHIFT) )
+	INPUT_PORT_DIGITAL_TYPE( 0, UI,      UI_FORCE_QUIT,       "UI Force Quit",          input_seq(KEYCODE_ESC, KEYCODE_LSHIFT) )
```
Comment (cuavas) [inline src/emu/inpttype.h]: I don’t like the choice of name.  “Force quit” implies that it prevents the system from preventing the quit from happening.  With this implementation, there’s no way for a device to cancel/defer the exit anyway, so it’s always forced.  It’s actually “quit without warning and without saving NVRAM contents”.

Comment (ajrhacker) [reply src/emu/inpttype.h]: Could you suggest alternate names for the input? I agree that its present nomenclature is a bit dubious, as are the finer aspects of its technical implementation.

Comment (cuavas) [reply src/emu/inpttype.h]: Read the overall comments at the top.  This input is something more along the lines of "abandon state" which requires a lot more thought to implement properly.

**Comment by ajrhacker** [comment]
The length of the feedback is promising. Though quite a bit of it is rather tangential to what I had in mind, there are a fair number of salient points to address here.

> A fixed exit delay isn’t going to help much when the required delay depends a lot on the system and the state it’s in.

This is true in principle, though I did make the delay a configurable option, and the default one-twentieth of a second should be close enough to the window a typical system running on AC power might expect. Of course, not all systems run on AC power, and a handheld device based on some sort of CMOS SoC handles power management very differently from a 1970s-style computer system with an industrial-strength backplane distributing regulated and unregulated voltages and half-rectified mains oscillations from a dedicated power supply board. To put it simply, I believe that no one solution to this problem is going to be perfect for all systems, so the "short-sighted" remark feels a bit unfair.

> I think a more useful upfront feature would be to make it possible for a device to notify that it’s entered soft power-down. The machine can subscribe to this notification on the root device, and exit the emulation session without additional interaction. Then, for example, when you hit the power button on the V.Smile or key 2nd, Off on the TI-83, you’ll get dropped back to the menu when the system’s ready.

I thought about implementing notifiers of this sort, along with having the devices implementing `device_power_fail` (or whatever it would be called) provide worst-case timeout periods. This seems more obviously suited to small CMOS handhelds than the old-style bus-based computer systems I was thinking of, where the individual cards might not have any signal to indicate that they have successfully powered down. (The Amusco hardware, despite using NMOS components, does set a few PPI output lines to signal that the NMI handler is done and the machine has nothing left to do except spin endlessly.)

There are a few good reasons I see for not being as enthusiastic as you are about implementing an automatic exit condition:
* If there is more than just one CPU and each CPU can go into soft power-down mode separately, which one should stop the emulation? The problem gets trickier in larger, modular systems, especially those implementing sophisticated power-management features allowing software to power down unused subsystems or the system management processor to power-cycle the main CPU. The whole idea of making the root device special seems to be a legacy that MAME is slowly and painfully moving away from.
* Some handheld systems, which seem to include the ST2XXX-based VTech laptops, will enter power-down mode soon after initializing the processor, requiring the user to actively manipulate the power button to wake them up. This obviously should not trigger an exit from emulation.
* Handheld systems also have a tendency to automatically power themselves down if left untouched for a while to conserve battery power. (GameKing III seems to do this.) Many users unfamiliar with these systems will not expect the machine or the application to quit itself without their input, and more experienced users would probably rather try to power them back up at that point.

As a comparative note, MAME's "UI Soft Reset" input is another universally imposed abstraction that obscures a wide variation in how reset circuits are technically implemented. It's inaccurate enough for a few systems that they have to provide their own custom "reset" inputs that work more selectively. It probably still serves a greater percentage of machines adequately enough than this proposal does, but one size still doesn't fit all. Even if some form of `device_power_fail` is implemented, power-off inputs will need to remain separate inputs on systems where those are filtered through GPIO or keyboard connectors rather than being direct power supply cutoffs.

As a side note, `emu/uiinput.cpp` is an atrocious kludge barely good enough to process most of the input types involved here. It really doesn't help that exiting any machine, including temporarily during a hard reset, tears down the whole core input system.

> What about cartridge NVRAM (“battery save” in device_image_interface)?

I specifically took care of that in `diimage.cpp` after failing to do so when I originally implemented `nvram_save`.

> What about disks? Abandoning changes to NVRAM without being able to roll back disk changes could cause problems for some machines.

It's not clear what kind of machines you're referring to here.

It might be a good idea, even independent of this feature, to have a UI info screen indicate CPUs that have successfully powered themselves down.

File: src/emu/inpttype.h (hunk 1 above)
Comment (ajrhacker) [reply src/emu/inpttype.h]: I don't take that as a meaningful answer, though I'm not in any hurry to merge this PR now and may leave out the "Force Quit" part when I revise it later.

**Comment by happppp** [comment]
Related: https://github.com/mamedev/mame/issues/728

I like it, I don't think it needs to be perfect right from the start. It can be improved when necessary. Of course there will always be edge-cases, like the 2-button press for TI83, or a "Are you sure?" from the system after OFF button like on risc2500.

One idea I'd have is to make the delay a 'max delay' and allow the driver to tell the MAME core to say it's ok to power-off now (MAME tells driver it's exiting, max 1 sec -> driver handles power-down -> 0.08 sec later, driver tells MAME it finished -> MAME exits). 1/20th of a second is not enough for systems that can only power down with a button, like Neo Geo Pocket.
If the delay expired, this would also allow MAME to warn the user that power-down failed.

Optionally auto-exiting MAME after the user powers down the emulated system with a button(s) is not the same feature as this. But it's also a thing we can add later.

**Comment by smf-** [comment]
On 19/11/2019 11:28, hap wrote:
> I like it, I don't think it needs to be perfect right from the start.
> It can be improved when necessary. Of course there will always be
> edge-cases, like the 2-button press for TI83, or a "Are you sure?"
> from the system after OFF button like on risc2500.

The edge cases are more likely to be what happens when we have multiple
devices running in one mame context, or something on a slot device that
needs to be powered off.

The wrong implementation could be really halmful to long term development.

**Comment by happppp** [comment]
I don't think multiple devices will be an edge case. With the max-delay idea, MAME would need to exit after all devices tell it's ok to exit.

**Comment by smf-** [comment]
If you have two computers linked together and one shuts down first, then
only the second one should continue executing.

**Comment by happppp** [comment]
I was mainly talking about the user exiting MAME. In this case: user exits MAME -> MAME tells both emulated computers that a shutdown is pending -> drivers do power-off procedure if needed.

**Comment by happppp** [comment]
@ajrhacker what do you think about max-delay instead of fixed delay, plus extra callback?

For example this is what I would have in mind for saitek_stratos.
See driver note here: https://github.com/mamedev/mame/blob/master/src/mame/drivers/saitek_stratos.cpp#L16

What I would do here, when user exits MAME and a power-off pending callback is added, is simulating a STOP button press. Then over here: https://github.com/mamedev/mame/blob/master/src/mame/drivers/saitek_stratos.cpp#L215 let MAME know that the power-off is finished so MAME can exit.

**Comment by smf-** [comment]
Setting time outs is going to be difficult because it will depend on the
piece of software you are running.

I would do something like

1. Add support for power buttons that can shut the computer down/start it up

2. Having escape trigger any power buttons and wait for everything to
shut down

3. Hold escape for ten seconds to force exit

**Comment by ajrhacker** [comment]
I'll close this, having not much interest right now in redoing this in a way that pleases most people, but the discussion was worth it.

I think the best way to handle battery-operated devices would be to notify the user that the CPU has turned itself off, with an optional plugin that automatically exits emulation in case of that event. On some systems, the sleeping CPU can still be revived by a button press that causes an interrupt, so a soft reset might not be necessary.

**Pull request closed without merging**